### PR TITLE
Support List and LargeList in Row format (#3159)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #3159

# Rationale for this change

The longer term goal is to make the Row Format support enough types to use in DataFusion so we can use it for a unified GroupBy operation 
 

# What changes are included in this PR?

Add support for encoding/decoding lists from the row format

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/row/mod.rs [hunk 1]
```diff
@@ -343,6 +344,54 @@ mod variable;
...
+///     ┌──┬──┐     ┌──┬──┐     ┌──┬──┐     ┌──┬──┐        ┌──┬──┐
+///  1  │01│01│  2  │01│02│  3  │01│02│  1  │01│01│  null  │00│00│
+///     └──┴──┘     └──┴──┘     └──┴──┘     └──┴──┘        └──┴──┘
+///```
+///
+/// Which would be grouped into the following canonical byte arrays:
+///
+/// ```text
+///                         ┌──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┐
+///  [1_u8, 2_u8, 3_u8]     │01│01│01│02│01│03│00│00│00│02│00│00│00│02│00│00│00│02│00│00│00│03│
```
Comment (crepererum) [inline arrow/src/row/mod.rs:383]: I get the `│01│01│01│02│01│03│` prefix and the `│00│00│00│03│` suffix. But where do the other bytes come from?

File: arrow/src/row/mod.rs [hunk 2]
```diff
@@ -343,6 +344,54 @@ mod variable;
...
+/// [1_u8, null]
+/// []
+/// null
+/// ```
+///
+/// The elements would be converted to:
+///
+/// ```text
+///     ┌──┬──┐     ┌──┬──┐     ┌──┬──┐     ┌──┬──┐        ┌──┬──┐
+///  1  │01│01│  2  │01│02│  3  │01│02│  1  │01│01│  null  │00│00│
```
Comment (crepererum) [inline arrow/src/row/mod.rs:375]:
```suggestion
///  1  │01│01│  2  │01│02│  3  │01│03│  1  │01│01│  null  │00│00│
```

File: arrow/src/row/mod.rs (hunk 1 above)
Comment (tustvold) [reply arrow/src/row/mod.rs:383]: > **the lengths of each encoded row**, and the number of elements

Comment (alamb) [reply arrow/src/row/mod.rs:383]: So basically `00│00│00│02│00│00│00│02│00│00│00│02│00│00│00│03` represents `[element0_len, element1_len, element2_len, element count]` --> `[2, 2, 2, 3]`

File: arrow/src/row/mod.rs [hunk 3]
```diff
@@ -381,13 +430,29 @@ enum Codec {
...
+                // The encoded contents will be inverted if descending is set to true
+                // As such we set `descending` to false and negate nulls first if it
+                // it set to true
+                let options = SortOptions {
+                    descending: false,
+                    nulls_first: sort_field.options.nulls_first
+                        != sort_field.options.descending,
+                };
+
+                let field = SortField::new_with_options(f.data_type().clone(), options);
```
Comment (alamb) [inline arrow/src/row/mod.rs:452]: I don't understand how descending sort is achieved if the list is always encoded descending false 🤔 )

However, I see it is tested below, so 👍 

```rust
       let options = SortOptions {
            descending: true,
            nulls_first: false,
        };
```

File: arrow/src/row/mod.rs [hunk 4]
```diff
@@ -1671,6 +1796,200 @@ mod tests {
...
+        builder.values().append_value(32);
+        builder.append(true);
+        builder.values().append_value(32);
+        builder.values().append_value(52);
+        builder.values().append_value(12);
+        builder.append(true);
+        builder.values().append_value(32);
+        builder.values().append_value(52);
+        builder.append(true);
+        builder.values().append_value(32); // MASKED
```
Comment (alamb) [inline arrow/src/row/mod.rs:1812]: masked means this row is NULL so these values should be igored

File: arrow/src/row/mod.rs [hunk 5]
```diff
@@ -1671,6 +1796,200 @@ mod tests {
...
+            descending: true,
+            nulls_first: true,
+        };
+        let field = SortField::new_with_options(d, options);
+        let mut converter = RowConverter::new(vec![field]).unwrap();
+        let rows = converter.convert_columns(&[Arc::clone(&list)]).unwrap();
+
+        assert!(rows.row(0) < rows.row(1)); // [32, 52, 32] < [32, 52, 12]
+        assert!(rows.row(2) > rows.row(1)); // [32, 42] > [32, 52, 12]
+        assert!(rows.row(3) < rows.row(2)); // null < [32, 42]
```
Comment (alamb) [inline arrow/src/row/mod.rs:1888]: 👍 

Thank you for the comments -- they make the tests easy to follow

File: arrow/src/row/mod.rs [hunk 6]
```diff
@@ -1671,6 +1796,200 @@ mod tests {
...
+        // [
+        //   [[1, 2], [1, null]],
+        //   [[1, null], [1, null]],
+        //   [[1, null], null]
+        //   null
+        //   [[1, 2]]
+        // ]
+        let options = SortOptions {
+            descending: false,
+            nulls_first: true,
```
Comment (alamb) [inline arrow/src/row/mod.rs:1944]: I recommend adding a test in nested lists for `nulls_first: false`, and verify that 

```rust
        assert!(rows.row(0) < rows.row(1));
```

**Review by alamb — APPROVED** [review]
I went through the code and tests carefully. It is a little mind bending but I think it is very nicely done 🏆

File: arrow/src/row/mod.rs (hunk 3 above)
Comment (tustvold) [reply arrow/src/row/mod.rs:452]: Only the elements are encoded with descending false, they are then encoded using variable length encoding which may reorder them. Yes it is mind-bending :sweat_smile:

File: arrow/src/row/mod.rs (hunk 1 above)
Comment (crepererum) [reply arrow/src/row/mod.rs:383]: maybe we can add this explanation to the docstring?

**Comment by alamb** [comment]
I can feel DataFusion grouping getting faster..... 

high-five

**Pull request merged**